libweighted_tree/print_hierarchy.cpp

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2020 Lawrence Livermore National Security, LLC
+(c.f. AUTHORS, NOTICE.LLNS, COPYING)
+
+This file is part of the Flux resource manager framework.
+For details, see https://github.com/flux-framework.
+
+SPDX-License-Identifier: LGPL-3.0
+*/
+#include <iostream>
+#include <sqlite3.h>
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <vector>
+#include <tuple>
+
+std::string hierarchy = "Bank|User|RawShares\n";
+std::string indent = "";
+char* messageError;
+sqlite3_stmt *res;
+
+int get_sub_banks(sqlite3* DB, const std::string& bank_name, const std::string& indent) {
+  int exit = 0;
+
+  // get shares of the current bank
+  std::string select_shares_stmt = "SELECT bank_table.shares "
+          "FROM bank_table "
+          "WHERE bank='";
+  select_shares_stmt.append(bank_name + "'");
+  exit = sqlite3_prepare_v2(DB, select_shares_stmt.c_str(), -1, &res, 0);
```

Comment:
I'm not too familiar with the sqlite3 API, but it looks like from the documentation that you can actually `prepare`/compile the statement with `?`s (just like with pandas) and then use the `bind` function to apply specific parameters to the prepared statement.  Prepared Statement documentation w/ the lifecycle: https://sqlite.org/c3ref/stmt.html

I think this has two implications for the code:

1. You will only need to prepare the statement one time, but it is bound and reset multiple times.
2. You will need to replace the statment append with a call to a `sqlite3_bind_*()` function

For example:
```C++
  // put this in the main function
  std::string select_shares_stmt = "SELECT bank_table.shares "
          "FROM bank_table "
          "WHERE bank='?'";
  sqlite3_stmt select_bank_stmt;
  exit = sqlite3_prepare_v2(DB, select_shares_stmt.c_str(), -1, &select_bank_stmt, 0);
  
  // keep this in the get_sub_banks function
  exit = sqlite3_bind_text(&select_bank_stmt, 0, bank_name.c_str(), -1, NULL);
  exit = sqlite3_step(select_bank_stmt);
  std::string bank_shares = ... sqlite3_column_text(...);
  sqlite3_reset(&select_shares_stmt);
```

Reply:
Thanks @SteVwonder! This was super useful feedback. I believe I was able to get this working, and I pushed those changes up. I noticed a couple of tiny subtleties but I was able to get the same results as before:

- I think parameters are automatically quoted and escaped, so we don't actually need to surround the parameter with quotes. I referred to this GitHub thread https://github.com/mapbox/node-sqlite3/issues/999:

```c
std::string select_shares_stmt = "SELECT bank_table.shares FROM bank_table WHERE bank=?";
```

- The second parameter to `sqlite3_bind_text()` is the index of the SQL parameter to be set, and I guess the leftmost index is 1 instead of 0. Reference: https://sqlite.org/c3ref/bind_blob.html

Let me know what you think of these changes. 

I think ultimately this PR should wait until @dongahn's weighted tree-based fairness PR gets landed to merge this one, so I'll keep it as a draft for now. 🙂